### PR TITLE
Use Sonata admin tabs

### LINKED ISSUE #42: Move translation catalogues to this bundle
Currently, all translation catalogues still live in the core bundles. They should be moved to this bundle.

We may have to split it in some catalogues, to avoid conflicts.

### PR DESCRIPTION
**Pull request by wouterj**
fix #42, #27, #4 

Todo
---

 * [x] Use tabs in the block bundle
 * [x] Translate SeoBundle fields (problems with Form component integration)
 * [x] Make form_group and form_tab of the extensions configurable

### CODE REVIEW

**Comment by wouterj** [comment]
/ping @dbu @ElectricMaxxx Ready to review & merge imo

**Comment by wouterj** [comment]
@ElectricMaxxx can you please add me maintainer as maintainer on packagist or enable the packagist service for this repository yourself? If you've done so, using `dev-issue-4/use-tabs` on your sandbox branch will give you a demo of this PR.

**Comment by ElectricMaxxx** [comment]
I hope i did both now. So you are maintainer and i hopefully activated the hook.

**Comment by ElectricMaxxx** [comment]
Menu:
auswahl_064

Route:
auswahl_063

Content:
auswahl_062

My Opinion:
Awesome

**Comment by wouterj** [comment]
Please also note that most extensions now have dedicated tabs on the page. These can be customized in the settings (e.g. one can configure to show the publish workflow stuff on the main tab as well).

**Comment by ElectricMaxxx** [comment]
@wouterj yes found it. Really nice now. Never thought, that we can combine blocks and tabs :-)

File: src/Admin/Block/AbstractBlockAdmin.php [hunk 1]
```diff
@@ -21,7 +24,7 @@
     /**
      * @var string
      */
-    protected $translationDomain = 'CmfBlockBundle';
+    protected $translationDomain = 'CmfSonataAdminIntegrationBundle';
```
Comment (dbu) [inline src/Admin/Block/AbstractBlockAdmin.php]: technically we could, but i think its not needed. unless we do one catalog per topic, people can just overwrite the CmfSonataAdminIntegrationBundle catalog in app/Resources. i think that is good enough.

File: src/Admin/Block/ContainerBlockAdmin.php [hunk 2]
```diff
@@ -38,12 +36,7 @@ protected function configureListFields(ListMapper $listMapper)
      */
     protected function configureFormFields(FormMapper $formMapper)
     {
-        $formMapper
-            ->with('form.group_general')
-                ->add('parentDocument', TreeModelType::class, array('root_node' => $this->getRootPath(), 'choice_list' => array(), 'select_root_node' => true))
-                ->add('name', TextType::class)
-            ->end()
-        ;
+        parent::configureFormFields($formMapper);
```
Comment (dbu) [inline src/Admin/Block/ContainerBlockAdmin.php]: if the only thing we do is call the parent, lets just remove the method completely here

File: src/Admin/Block/Imagine/SlideshowBlockAdmin.php [hunk 3]
```diff
@@ -58,32 +57,28 @@ protected function configureListFields(ListMapper $listMapper)
...
-                ->add('children', CollectionType::class,
-                    array(),
-                    array(
-                        'edit' => 'inline',
-                        'inline' => 'table',
-                        'admin_code' => $this->embeddedAdminCode,
-                        'sortable' => 'position',
-                    ))
+            ->tab('form.tab_general')
+                ->with('form.group_block', null === $this->getParentFieldDescription()
```
Comment (dbu) [inline src/Admin/Block/Imagine/SlideshowBlockAdmin.php]: is this ever null? we call parent above, so i would expect this to always have something.

File: src/Admin/Core/Extension/PublishableExtension.php [hunk 4]
```diff
@@ -27,25 +27,41 @@ class PublishableExtension extends AbstractAdminExtension
...
+        if ($formMapper->hasOpenTab()) {
+            $formMapper->end();
+        }
+
         $formMapper
-            ->with($this->formGroup, array('translation_domain' => 'CmfCoreBundle'))
-                ->add('publishable', CheckboxType::class, array(
-                    'required' => false,
-                ))
+            ->tab($this->formTab, 'form.tab_publish' === $this->formTab
```
Comment (dbu) [inline src/Admin/Core/Extension/PublishableExtension.php]: what if i configure the publish to be in the form.tab_default (or whatever the default tab is called)? will this overwrite the tab, or add to the tab? which options win, the first or the last? or are they merged?

File: src/Admin/Routing/Extension/RouteReferrersExtension.php [hunk 5]
```diff
@@ -23,21 +23,34 @@
...
-            ->add(
-                'routes',
-                CollectionType::class,
-                array(),
-                array(
-                    'edit' => 'inline',
-                    'inline' => 'table',
-                )
-            )
+            ->tab($this->formTab)
```
Comment (dbu) [inline src/Admin/Routing/Extension/RouteReferrersExtension.php]: don't we need the translation_domain here as well?

File: src/Admin/Routing/RedirectRouteAdmin.php [hunk 6]
```diff
@@ -38,12 +38,23 @@ protected function configureListFields(ListMapper $listMapper)
     protected function configureFormFields(FormMapper $formMapper)
     {
         $formMapper
-            ->with('form.group_general')
-                ->add('parentDocument', TreeModelType::class, array('choice_list' => array(), 'select_root_node' => true, 'root_node' => $this->routeRoot))
+            ->with('form.group_location', ['class' => 'col-md-3'])
```
Comment (dbu) [inline src/Admin/Routing/RedirectRouteAdmin.php]: what about translation domain?

File: src/Admin/Routing/RouteAdmin.php [hunk 7]
```diff
@@ -56,42 +49,52 @@ protected function configureListFields(ListMapper $listMapper)
...
-                    'translation_domain' => 'CmfRoutingBundle',
-                ))
-                    ->add('content', TreeModelType::class, array('choice_list' => array(), 'required' => false, 'root_node' => $this->contentRoot))
-                ->end()
-                ->with('form.group_advanced', array(
-                    'translation_domain' => 'CmfRoutingBundle',
-                ))
-                    ->add('variablePattern', TextType::class, array('required' => false), array('help' => 'form.help_variable_pattern'))
+            ->tab('form.tab_general')
+                ->with('form.group_location', ['class' => 'col-md-3'])
```
Comment (dbu) [inline src/Admin/Routing/RouteAdmin.php]: translation domain here and above?

File: src/Admin/Routing/RouteAdmin.php [hunk 8]
```diff
@@ -128,17 +131,18 @@ public function getExportFormats()
     protected function configureFieldsForDefaults($dynamicDefaults)
     {
         $defaults = array(
-            '_controller' => array('_controller', TextType::class, array('required' => false)),
-            '_template' => array('_template', TextType::class, array('required' => false)),
-            'type' => array('type', RouteTypeType::class, array(
+            '_controller' => ['_controller', TextType::class, ['required' => false, 'translation_domain' => 'CmfSonataAdminIntegrationBundle']],
+            '_template' => ['_template', TextType::class, ['required' => false, 'translation_domain' => 'CmfSonataAdminIntegrationBundle']],
```
Comment (dbu) [inline src/Admin/Routing/RouteAdmin.php]: this is no admin extension, why specify the translation domain on a per field level?

File: src/DependencyInjection/Factory/CoreAdminFactory.php [hunk 9]
```diff
@@ -34,26 +34,21 @@ public function getKey()
...
+                            ->scalarNode('form_tab')->defaultValue('form.tab_publish')->end()
                         ->end()
                     ->end()
                     ->arrayNode('publish_time')
                         ->addDefaultsIfNotSet()
                         ->children()
                             ->scalarNode('form_group')->defaultValue('form.group_publish_workflow')->end()
-                        ->end()
-                    ->end()
-                    ->arrayNode('translatable')
```
Comment (dbu) [inline src/DependencyInjection/Factory/CoreAdminFactory.php]: what about translatable? or is this now completely handled by SonataTranslationBundle?

File: tests/Unit/Admin/Core/Extension/PublishTimePeriodExtensionTest.php [hunk 10]
```diff
@@ -1,39 +0,0 @@
...
- *
- * For the full copyright and license information, please view the LICENSE
- * file that was distributed with this source code.
- */
-
-namespace Symfony\Cmf\Bundle\SonataAdminIntegrationBundle\Tests\Unit\Admin\Core\Extension;
-
-use Symfony\Cmf\Bundle\SonataAdminIntegrationBundle\Admin\Core\Extension\PublishTimePeriodExtension;
-
-class PublishTimePeriodExtensionTest extends \PHPUnit_Framework_TestCase
```
Comment (dbu) [inline tests/Unit/Admin/Core/Extension/PublishTimePeriodExtensionTest.php]: why do you drop these tests?

File: src/Admin/Block/Extension/BlockCacheExtension.php [hunk 11]
```diff
@@ -22,16 +22,37 @@
...
-                ->add('ttl', TextType::class)
+            ->tab($this->formTab, 'form.tab_general' === $this->formtab
+                ? ['translation_domain' => 'CmfSonataAdminIntegrationBundle']
+                : []
+            )
+                ->with($this->formGroup, 'form.group_metadata' === $this->formGroup
+                    ? ['translation_domain' => 'CmfSonataAdminIntegrationBundle']
+                    : []
+                )
+                    ->add('ttl', TextType::class)
```
Comment (dbu) [inline src/Admin/Block/Extension/BlockCacheExtension.php]: as you only set the translation domain conditionally on the tab + group, you should set the domain on every field in the admin extensions. can you check them again and add that please?

File: src/Admin/Block/ContainerBlockAdmin.php (hunk 2 above)
Comment (wouterj) [inline src/Admin/Block/ContainerBlockAdmin.php]: Yeah, see #45 . This admin is missing the `children` field, making it pretty useless at the moment. Will fix this in another PR

File: src/Admin/Block/Imagine/SlideshowBlockAdmin.php (hunk 3 above)
Comment (wouterj) [inline src/Admin/Block/Imagine/SlideshowBlockAdmin.php]: To be honest, I have no idea what the `getParentFieldDescription()` method is doing and why we commonly compare it against null. I just copy/pasted it from the original code.

File: src/Admin/Core/Extension/PublishableExtension.php (hunk 4 above)
Comment (wouterj) [inline src/Admin/Core/Extension/PublishableExtension.php]: It'll add to the tab and options will be merged

File: src/Admin/Routing/Extension/RouteReferrersExtension.php (hunk 5 above)
Comment (wouterj) [inline src/Admin/Routing/Extension/RouteReferrersExtension.php]: No, as this tab is already configured in the admin against the correct translation domain.

File: src/Admin/Routing/RedirectRouteAdmin.php (hunk 6 above)
Comment (wouterj) [inline src/Admin/Routing/RedirectRouteAdmin.php]: Not needed, if the field has no translation domain, the domain of the admin will be used.

File: src/Admin/Routing/RouteAdmin.php (hunk 7 above)
Comment (wouterj) [inline src/Admin/Routing/RouteAdmin.php]: same

File: src/Admin/Routing/RouteAdmin.php (hunk 8 above)
Comment (wouterj) [inline src/Admin/Routing/RouteAdmin.php]: For some reason, this options field thing requires us to set the translation domain each time.

File: src/DependencyInjection/Factory/CoreAdminFactory.php (hunk 9 above)
Comment (wouterj) [inline src/DependencyInjection/Factory/CoreAdminFactory.php]: Indeed, the extension was removed in CMF 2.0

File: tests/Unit/Admin/Core/Extension/PublishTimePeriodExtensionTest.php (hunk 10 above)
Comment (wouterj) [inline tests/Unit/Admin/Core/Extension/PublishTimePeriodExtensionTest.php]: They literally test nothing and were harder to write as there are now chained tab with calls, etc.

**Comment by dbu** [comment]
ah, i think it might be a roundabout way to see if the document we are
processing is new or existing?

File: src/Admin/Routing/RouteAdmin.php (hunk 8 above)
Comment (dbu) [inline src/Admin/Routing/RouteAdmin.php]: we could add a comment to not confuse our future selves.

**Comment by dbu** [comment]
thanks for the clarifications. this looks good to me then. you marked the PR WIP, but i think we can just merge it then and further clean up in separate PRs.

**Comment by ElectricMaxxx** [comment]
Same from. Me found some missing translations through the debug toolbar in sandbox (there is a branch checked out from my branch with this work inside) But will collect them in a separate issue.

**Comment by wouterj** [comment]
Merging

File: src/Resources/translations/CmfSonataAdminIntegrationBundle.en.xliff [hunk 12]
```diff
@@ -16,6 +16,15 @@
...
             </trans-unit>
 
+            <trans-unit id="filter.label_title">
+                <source>filter.label_title</source>
+                <target>Title</target>
+            </trans-unit>
+            <trans-unit id="filter.label_name">
+                <source>filter.label_name</source>
+                <target>Name</target>
+            </trans-unit>
```
Comment (wouterj) [inline src/Resources/translations/CmfSonataAdminIntegrationBundle.en.xliff]: these are missing in the other translation files, we need to add them there

**Pull request merged**